compiler/next/include/chpl/Queries/Context.h

```diff
@@ -0,0 +1,352 @@
+conceptually a vector of top-level symbol), the ``combine`` function
+should try to update only those elements of ``keep`` that changed by swapping
+in the appropriate elements from ``addin``. This strategy allows later queries
+that depend on such a result to use pointers to the owned elements and to
+avoid updating everything if just one element changed.
+
+Queries *can* return results that contain non-owning pointers to results from
+dependent queries. In that event, the combine function should not rely on the
+contents of these pointers. The system will make sure that they refer to valid
+memory but they might be a combination of old results. Additionally, the system
+will ensure that any old results being replaced will remain allocated until the
+garbage collection runs outside of any query.
+
+For example, a ``parse`` query might result in a list of ``owned`` AST element
+pointers. A follow-on ``listSymbols`` result in something containing these AST
+element pointers, but not owning them. The ``listSymbols`` query needs to use a
+``combine`` function that does not look into these queries.
+
+\endrst
+
+ */
+class Context {
+ private:
+  // map that supports uniqueCString / UniqueString
+   typedef std::unordered_map<const char*, char*, chpl::detail::UniqueStrHash, chpl::detail::UniqueStrEqual> UniqueStringsTableType;
+  UniqueStringsTableType uniqueStringsTable;
+
+  // map from a query name to appropriate QueryMap object.
+  // maps to an 'owned' heap-allocated thing to manage having subclasses
+  // without slicing.
+  std::unordered_map<UniqueString, owned<chpl::querydetail::QueryMapBase>> queryDB;
+
+  // Since IDs include module names but not file paths, use this
+  // map to go from module name to file path.
+  // (If this proves too restrictive for some reason, we could
+  //  start including file path in IDs).
+  std::unordered_map<UniqueString, UniqueString> modNameToFilepath;
+
+  struct QueryDepsEntry {
+    UniqueString queryName;
+    chpl::querydetail::QueryDependencyVec dependencies;
+    std::vector<ErrorMessage> errors;
+    QueryDepsEntry(UniqueString queryName)
+      : queryName(queryName), dependencies(), errors() {
+    }
+  };
+
+  // this is used to compute the dependencies
+  std::vector<QueryDepsEntry> queryDeps;
+
+  chpl::querydetail::RevisionNumber currentRevisionNumber;
+  chpl::querydetail::RevisionNumber lastPrepareToGCRevisionNumber;
+  chpl::querydetail::RevisionNumber gcCounter;
+
+  Context();
+  const char* getOrCreateUniqueString(const char* s);
+  bool queryCanUseSavedResult(chpl::querydetail::QueryMapResultBase* resultEntry);
+  void saveDependenciesAndErrorsInParent(chpl::querydetail::QueryMapResultBase* resultEntry);
+  void endQueryHandleDependency(chpl::querydetail::QueryMapResultBase* result);
+
+  template<typename ResultType, typename... ArgTs>
+  chpl::querydetail::QueryMapResult<ResultType>*
+  updateResultForQuery(
+      const std::tuple<ArgTs...>& tupleOfArgs,
+      ResultType result,
```

Comment:
For precedence in handling C++ tuples, see: https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/library/src/include/tuple_helper.hpp